asf_heat_pump_suitability/pipeline/prepare_features/listed_buildings.py

```diff
@@ -0,0 +1,84 @@
+import polars as pl
+import geopandas as gpd
+import pandas as pd
+from asf_heat_pump_suitability.getters.get_datasets import (
+    get_df_historicengland_listedbuildings,
+)
+
+
+def transform_df_EPC_X_and_Y_to_point(
+    enhanced_epc_df: pl.DataFrame,
+    x_col: str = "X_COORDINATE",
+    y_col: str = "Y_COORDINATE",
+) -> gpd.GeoDataFrame:
+    """
+    Transform 'X' and 'Y' coordinates in EPC dataset to be in 'POINT' format.
+
+    Args:
+        df (pl.DataFrame): Enhanced EPC dataset
+
+    Returns:
+        pl.DataFrame: EPC dataset with a column of 'POINT' (transformed from 'X' and 'Y' coordinate).
+    """
+    enhanced_epc_df = enhanced_epc_df.to_pandas()
+    gdf = gpd.GeoDataFrame(
+        enhanced_epc_df,
+        geometry=gpd.points_from_xy(enhanced_epc_df[x_col], enhanced_epc_df[y_col]),
+        crs="EPSG:27700",
+    )
+    return gdf
+
+
+def get_filtered_df_listed_buildings() -> gpd.GeoDataFrame:
+    """
+    Get and filter out listed buildings from Historic England dataset.
+
+    Returns:
+        gpd.GeoDataFrame: Filtered Historic England dataset with only listed buildings grade and geometry.
+    """
+    historic_england_gdf = get_df_historicengland_listedbuildings()
+    relevant_columns = ["geometry", "Grade"]
+    filtered_historic_england_gdf = historic_england_gdf[relevant_columns]
+
+    return filtered_historic_england_gdf
+
+
+def spatial_join_epc_with_listed_buildings(
+    enhanced_epc_df: pl.DataFrame, listed_buildings_df: gpd.GeoDataFrame
+) -> gpd.GeoDataFrame:
```

Comment:
You can stop the memory issues and speed things up by joining the EPC data to the listed building dataset in chunks. I modified this function to do this and it took 3 minutes to process the 16 million properties in `s3://asf-heat-pump-suitability/outputs/2023_Q2_EPC_enhanced_weights.parquet`. A few tweaks were needed to what you had before (e.g. the inner join on `gpd.sjoin`):


```
from tqdm import tqdm
import logging

def spatial_join_epc_with_listed_buildings(
    enhanced_epc_df: pl.DataFrame, listed_buildings_df: gpd.GeoDataFrame
) -> pl.DataFrame:
    """
    Spatial join EPC dataset with listed buildings dataset.
    Args:
        enhanced_epc_df (pl.DataFrame): Enhanced EPC dataset
        listed_buildings_df (gpd.GeoDataFrame): Filtered Historic England dataset with only listed buildings grade and geometry.
    Returns:
        pl.DataFrame: EPC dataset with listed buildings grade without geometry.
    """

    # Add a unique index column for merging later
    enhanced_epc_df = enhanced_epc_df.with_row_index("index")

    chunk_size = 100000
    partitions = enhanced_epc_df.with_row_count("chunk_id").select(pl.col("chunk_id") // chunk_size)
    # group based on the created index, resulting chunk_size partitons
    data_partitioned = enhanced_epc_df.with_columns(partitions).partition_by("chunk_id")
    logging.info(f"Adding listed buildings to EPC in {len(data_partitioned)} chunks")
    for i, enhanced_epc_df_chunk in tqdm(enumerate(data_partitioned)):
        epc_gdf = transform_df_EPC_X_and_Y_to_point(enhanced_epc_df_chunk)
        epc_gdf_temp = epc_gdf[["geometry", "index"]].copy()
        joined_gdf_chunk = gpd.sjoin(
            epc_gdf_temp, listed_buildings_df, how="inner", predicate="intersects"
        )
        # Drop the geometry column
        joined_gdf_chunk = joined_gdf_chunk.drop(columns=["geometry", "index_right"])
        if i==0:
            joined_gdf = joined_gdf_chunk
        else:
            joined_gdf = pd.concat([joined_gdf, joined_gdf_chunk])

    enhanced_epc_df = enhanced_epc_df.join(pl.from_pandas(joined_gdf), how="left", on="index")
    
    return enhanced_epc_df
```

Reply:
(note the output is now a polars df so you probably won't need to apply `convert_gpd_to_polars` anymore)

Reply:
Did you know that the name tqdm stands for "taqaddum" in Arabic which can mean "progress"?  🚀

Reply:
Thanks for this, works a charm!

Reply:
haha that makes far more sense! 😆  I'm sure I read somewhere once that it meant "thank you demasiado" (i.e. thank you very much (spanish) for helping me see my progress), so have always thought it meant that - even though it makes **very** little sense!